### PR TITLE
chore: migration and testcases for 2factorauthentication files

### PR DESCRIPTION
**Pull request by utkarsha-deriv**
## Changes:

Please provide a summary of the change.

### Screenshots:

Please provide some screenshots of the change.

### CODE REVIEW

File: packages/account/src/Sections/Security/TwoFactorAuthentication/two-factor-authentication.tsx [hunk 1]
```diff
@@ -192,13 +194,7 @@ const TwoFactorAuthentication = observer(() => {
 
     return (
         <section className='two-factor'>
-            <div
-                className={classNames('two-factor__wrapper', {
-                    'two-factor__wrapper-dashboard': is_appstore,
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/two-factor-authentication.tsx]: Di we remove corresponding styles from scss files?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx [hunk 2]
```diff
@@ -1,57 +1,70 @@
...
-                    const { code, message } = enable_response.error;
-                    if (code === 'InvalidOTP')
-                        return { digit_code: localize("That's not the right code. Please try again.") };
-                    return { digit_code: message };
-                }
-            } else {
-                return { digit_code: localize("That's not the right code. Please try again.") };
+        } else if (formik_ref.current?.isValid) {
+            const totp_action = is_enabled ? 'disable' : 'enable';
+            enable_response = await WS.authorized.accountSecurity({
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx]: can we move this to a hook

Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx]: okay

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx [hunk 3]
```diff
@@ -1,57 +1,70 @@
 import React from 'react';
 import classNames from 'classnames';
-import { Formik, Form, Field } from 'formik';
+import { Formik, Form, Field, FormikHelpers, FieldProps, FormikProps } from 'formik';
 import { Input, Button } from '@deriv/components';
 import { getPropertyValue, WS } from '@deriv/shared';
 import { localize } from '@deriv/translations';
 
-const DigitForm = ({ is_enabled, setTwoFAStatus, setTwoFAChangedStatus, is_language_changing }) => {
+export type TDigitForm = {
```
Comment (shaheer-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx]: please dont export prop types.
you can use `React.ComponentProps<typeof DigitForm >` in test files or wherever you are importing this.

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx [hunk 4]
```diff
@@ -77,12 +90,11 @@ const DigitForm = ({ is_enabled, setTwoFAStatus, setTwoFAChangedStatus, is_langu
                                     value={values.digit_code}
                                     onChange={e => {
                                         handleChange(e);
-                                        setReadyForVerification(false);
                                     }}
                                     onBlur={handleBlur}
                                     required
                                     error={touched.digit_code && errors.digit_code}
-                                    maxLength='6'
+                                    maxLength={6}
```
Comment (shaheer-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx:94]: just a doubt. are we following this format everywhere?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx (hunk 3 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx]: Okay

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx (hunk 4 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx:94]: Yes, in quite some places

File: packages/account/src/Sections/Security/TwoFactorAuthentication/two-factor-authentication.tsx [hunk 5]
```diff
@@ -48,24 +42,28 @@ const TwoFactorAuthentication = observer(() => {
...
     }, [email_address]);
 
     const getDigitStatus = React.useCallback(async () => {
         const status_response = await getTwoFAStatus();
-        if (status_response.error) {
-            setErrorMessage(status_response.error.message);
+        // status_response can be boolean or an error object
+        if (typeof status_response !== 'boolean') {
+            setErrorMessage(status_response?.error?.message);
+            setLoading(false);
```
Comment (matin-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/two-factor-authentication.tsx]: Very nice 👍🏼

**Review by matin-deriv — APPROVED** [review]

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
+import React from 'react';
+import { render, screen, waitFor } from '@testing-library/react';
+import DigitForm, { TDigitForm } from '../digit-form';
+import userEvent from '@testing-library/user-event';
+import { WS } from '@deriv/shared';
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx]: lets reorder these imports as per import rules in deriv app.

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+        const digitInput = screen.getByRole('textbox');
+        const invalidOTP = '786789';
+
+        userEvent.type(digitInput, invalidOTP);
+
+        const submitButton = screen.getByRole('button', { name: /Enable/i });
+        userEvent.click(submitButton);
+
+        await waitFor(async () => {
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx]: why do we need async here?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx]: here too

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication-article.spec.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+        const two_factor_authentication_article = screen.getByText('Two-factor authentication (2FA)');
+        expect(two_factor_authentication_article).toBeInTheDocument();
+    });
+
+    it('should render article description properly', () => {
+        render(<TwoFactorAuthenticationArticle />);
+
+        const description = screen.getByText(
+            'Protect your account with 2FA. Each time you log in to your account, you will need to enter your password and an authentication code generated by a 2FA app on your smartphone.'
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication-article.spec.tsx]: lets use regex for text in test cases

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+    });
+
+    it('should render 2FA article component for dekstop when 2FA is disabled', async () => {
+        (isMobile as jest.Mock).mockReturnValue(false);
+        (isDesktop as jest.Mock).mockReturnValue(true);
+
+        renderComponent({ store_config: store });
+
+        await waitFor(() => {
+            // const article_component = screen.getByTestId('two_fa_article_mobile');
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx]: remove this commented code

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+    });
+
+    it('should render digitform component when 2FA is disabled', async () => {
+        renderComponent({ store_config: store });
+
+        await waitFor(() => {
+            const digitform = screen.getByTestId('digitform_2fa_disabled');
+            expect(digitform).toBeInTheDocument();
+        });
+    });
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx]: why do we need async await in all these test cases? could you please check this

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+    it('should render 2FA article component for mobile when 2FA is disabled', async () => {
+        (isMobile as jest.Mock).mockReturnValue(true);
+        (isDesktop as jest.Mock).mockReturnValue(false);
+
+        renderComponent({ store_config: store });
+
+        await waitFor(() => {
+            // const article_component = screen.getByTestId('two_fa_article_mobile');
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx]: lets remove commented codes

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+            ...store,
+            client: {
+                ...store.client,
+                getTwoFAStatus: jest.fn().mockResolvedValue(true),
+                has_enabled_two_fa: true,
+            },
+        };
+        renderComponent({ store_config: new_store_config });
+
+        await waitFor(() => {
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx]: do we need to await here

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx]: same here

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+        await waitFor(() => {
+            const error = screen.getByText(/Length of digit code must be 6 characters./i);
+            expect(error).toBeInTheDocument();
+        });
+    });
+
+    it('should display error if user types invalid OTP', async () => {
+        render(<DigitForm {...mock_props} />);
+        const digitInput = screen.getByRole('textbox');
+        const invalidOTP = '786789';
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx]: do we need this to assign to variable? Its not reused anywhere. cant we directly give to userEvent.type

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+
+        userEvent.type(digitInput, invalidOTP);
+
+        const submitButton = screen.getByRole('button', { name: /Enable/i });
+        userEvent.click(submitButton);
+
+        await waitFor(async () => {
+            const error = screen.queryByText(/That's not the right code. Please try again./i);
+            expect(error).not.toBeInTheDocument();
+        });
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx:200]: The description says "should display error" , but we are checking error not to be in the document.?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+    it('should display error if response error object does not contain error code', async () => {
+        WS.authorized.accountSecurity.mockResolvedValue({
+            success: { message: '', code: '' },
+        });
+
+        render(<DigitForm {...mock_props} />);
+
+        const digitInput = screen.getByRole('textbox');
+        const invalidOTP = '786789';
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx:190]: do we need this to assign to a variable . It is not reused anywhere.

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx (hunk 15 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx]: We can, I used it to make the code more readable

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx (hunk 7 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx]: queryByText is asynchronous in nature, usually we write await before screen.queryByText. However, we already mention await before waitFor(), so I used async before arrow function. Is it redundant?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx (hunk 17 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx:190]: same reason

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication-article.spec.tsx (hunk 9 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication-article.spec.tsx]: Okay

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx (hunk 11 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx]: Inside 2FA component everytime we render , useEffect runs 2 functions - generateQrCode and getDigitStatus which update states internally. We need waitFor() so that state updates showup otherwise testcases would fail. Additionally, since all this is async in nature I added async as the functions make api calls and state updates happen asynchronously.

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx (hunk 12 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx]: My bad. I'll remove them

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx (hunk 13 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx]: Yes, since getDigitStatus sets Loading to false..meaning we update states eveytime we render. Also the function is async thus, used await

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx (hunk 7 above)
Comment (amina-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx]: The async is added in the top level. we dont need async inside await. right?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx (hunk 16 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx:200]: My bad, the description should have been `should not display error if response object does not contain error object`

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx (hunk 8 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx]: Will fix this

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication-article.spec.tsx (hunk 9 above)
Comment (shaheer-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication-article.spec.tsx]: @amina-deriv regex can be weak, right?
`220% Leverage` passes
- the test `/20% Leverage/`, 
- the test `/220% leverage/i`
going forward, do we really need to use regex in test files?

File: packages/components/stories/icon/icons.js
Comment (amina-deriv) [inline packages/components/stories/icon/icons.js:1]: This file changes are unrelated to this card. we can revert the changes in this file.

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx [hunk 18]
```diff
@@ -1,64 +1,74 @@
...
+    digit_code: string;
+};
+
+const DigitForm = () => {
+    const { client, common } = useStore();
+    const { is_language_changing } = common;
+    const { has_enabled_two_fa, setTwoFAChangedStatus, setTwoFAStatus } = client;
+
+    const { is_TwoFA_enabled, error, isSuccess, sendUserOTP } = useSendUserOTP();
+    const button_text = has_enabled_two_fa ? localize('Disable') : localize('Enable');
```
Comment (shaheer-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx:19]: can this be `Localize` and passed to `Button` as `children`?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+        client: {
+            has_enabled_two_fa: false,
+            email_address: 'test@dev.com',
+            is_switching: false,
+            setTwoFAChangedStatus: jest.fn(),
+        },
+    });
+
+    const renderComponent = ({ store_config = store }) => {
+        return render(
```
Comment (shaheer-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-authentication.spec.tsx]: `return` keyword can be omitted?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+            setTwoFAStatus: jest.fn(),
+            setTwoFAChangedStatus: jest.fn(),
+        },
+        common: {
+            is_language_changing: false,
+        },
+    });
+
+    const renderComponent = ({ store_config = store }) => {
+        return render(
```
Comment (shaheer-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/digit-form.spec.tsx]: `return` keyword can be omitted?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-disabled.spec.tsx [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+    };
+
+    const store = mockStore({
+        client: {
+            has_enabled_two_fa: false,
+        },
+    });
+
+    const renderComponent = ({ store_config = store, mock = mock_props }) => {
+        return render(
```
Comment (shaheer-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/__tests__/two-factor-disabled.spec.tsx]: `return` keyword can be omitted?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx [hunk 22]
```diff
@@ -1,64 +1,74 @@
...
-    const [is_success, setSuccess] = React.useState(false);
-    const [is_ready_for_verification, setReadyForVerification] = React.useState(false);
-    const button_text = is_enabled ? localize('Disable') : localize('Enable');
-    const formik_ref = React.useRef();
-    let enable_response;
+type TDigitFormValues = {
+    digit_code: string;
+};
+
+const DigitForm = () => {
```
Comment (shaheer-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx]: do we need a store observer here?
`const DigitForm = observer(() => { ...`

File: packages/api/src/hooks/useAccountSecurity.ts [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+import useAuthorize from './useAuthorize';
+import useRequest from '../useRequest';
+
+type TAccountSecurityPayload = NonNullable<
+    NonNullable<Parameters<ReturnType<typeof useRequest<'account_security'>>['mutate']>[0]>['payload']
+>;
+
+const useAccountSecurity = () => {
+    const { isSuccess: success_auth } = useAuthorize();
+    const { mutate, ...rest } = useRequest('account_security');
```
Comment (shaheer-deriv) [inline packages/api/src/hooks/useAccountSecurity.ts:11]: `useRequest` is deprecated, and now `useMutation`.
https://deriv-group.slack.com/archives/C3AHMSM3P/p1695697505953689?thread_ts=1670994836.610179&cid=C3AHMSM3P

File: packages/account/src/Sections/Security/TwoFactorAuthentication/two-factor-enabled.tsx [hunk 24]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import { Icon, ThemedScrollbars, Text } from '@deriv/components';
+import DigitForm from './digit-form';
+import { isMobile } from '@deriv/shared';
+import { Localize } from '@deriv/translations';
+
+const TwoFactorEnabled = () => {
+    return (
+        <ThemedScrollbars is_bypassed={isMobile()} className='two-factor__scrollbars'>
```
Comment (shaheer-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/two-factor-enabled.tsx]: can we use `is_mobile` from `ui` store here?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/two-factor-disabled.tsx [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+    secret_key: string;
+    qr_secret_key: string;
+    is_loading_secret: boolean;
+};
+
+const TwoFactorDisabled = ({ secret_key, qr_secret_key, is_loading_secret }: TTwoFactorDisabled) => {
+    return (
+        <React.Fragment>
+            <ThemedScrollbars
+                is_bypassed={isMobile()}
```
Comment (shaheer-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/two-factor-disabled.tsx]: here too `is_mobile`?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx [hunk 26]
```diff
@@ -1,64 +1,74 @@
...
+        }
+    }, [error]);
+
+    React.useEffect(() => {
+        if (isSuccess) {
+            setTwoFAStatus(is_TwoFA_enabled);
+            setTwoFAChangedStatus(true);
+            formik_ref.current?.resetForm();
+        }
+        // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (amina-deriv) [inline packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx:54]: why do we disable this?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx (hunk 22 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx]: works fine without it as well. Shall I add it?

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx (hunk 18 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx:19]: Actually no because`TButtonCommonProps` of `Button` component takes only string type for text prop

File: packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx (hunk 26 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/digit-form.tsx:54]: Because ts requests to add setTwoFAStatus and setTwoFAChangedStatus also in dependencies but I feel they are not required.

File: packages/account/src/Sections/Security/TwoFactorAuthentication/two-factor-enabled.tsx (hunk 24 above)
Comment (utkarsha-deriv) [reply packages/account/src/Sections/Security/TwoFactorAuthentication/two-factor-enabled.tsx:9]: I think let's incorporate it once the latest changes for is_mobile are pushed to master?

File: packages/api/src/hooks/useAccountSecurity.ts (hunk 23 above)
Comment (utkarsha-deriv) [reply packages/api/src/hooks/useAccountSecurity.ts:11]: will need a pull from master for the feature branch for it

**Review by amina-deriv — APPROVED** [review]

**Review by shaheer-deriv — APPROVED** [review]
